src/WooCommerce.php

```diff
@@ -52,6 +52,143 @@ public static function woocommerce_get_settings_gallerya(array $settings): array
+    ]);
+    woocommerce_wp_select([
+      'id'          => '_' . Plugin::L10N . '_video_source',
+      'value'       => get_post_meta(get_the_ID(), '_' . Plugin::L10N . '_video_source', true),
+      'label'       => __('Video source', Plugin::L10N),
+      'options'     => array( '' => __('Please select', Plugin::L10N), 'vimeo' => 'Vimeo', 'youtube' => 'YouTube'),
+    ]);
+    echo '</div>';
+
+  }
+
+  /**
+   * Ensures new product are saved before updating its meta data.
+   *
+   * New products are still not saved when updated_post_meta hook is called.
+   * Since we can not check if the meta keys were changed before running
+   * our custom functions (see updateDeliveryTime and updateSalePercentage),
+   * we are forcing the post to be saved before updating the meta keys.
+   *
+   * @implements woocommerce_process_product_meta
+   */
+  public static function saveNewProductBeforeMetaUpdate($post_id) {
+    $product = wc_get_product($post_id);
+    $product->save();
+  }
+
+   /**
+   * Saves custom fields for simple products.
+   *
+   * @implements woocommerce_process_product_meta
+   */
+  public static function woocommerce_process_product_meta($post_id) {
+    $custom_fields = [
+      '_' . Plugin::L10N . '_video_id',
+      '_' . Plugin::L10N . '_video_source',
+    ];
+
+    foreach ($custom_fields as $field) {
+      if (isset($_POST[$field])) {
+        if (!is_array($_POST[$field]) && $_POST[$field]) {
+          update_post_meta($post_id, $field, $_POST[$field]);
+        }
+        else {
+          delete_post_meta($post_id, $field);
+        }
+      }
+    }
+
+    $custom_fields_checkbox = [
+    ];
+
+    foreach ($custom_fields_checkbox as $field) {
+      $value = isset($_POST[$field]) && !is_array($_POST[$field]) && wc_string_to_bool($_POST[$field]) ? 'yes' : 'no';
+      update_post_meta($post_id, $field, $value);
+    }
+  }
+
+  /**
+   * Replaces featured image with video in single product galleries.
+   *
+   * @implements woocommerce_single_product_image_thumbnail_html
+   */
+  public static function replaceFeaturedImageWithVideo($html, $thumbnail_id) {
+    global $product;
+
+    // Check to make sure we're only targeting the featured image.
+    if ($thumbnail_id === $product->get_image_id()) {
+      $video_id = get_post_meta($product->get_id(), '_' . Plugin::L10N . '_video_id', TRUE);
+      $video_source = get_post_meta($product->get_id(), '_' . Plugin::L10N . '_video_source', TRUE);
+
+      if ($video_id && $video_source) {
+        ob_start();
+
+        $gallery_thumbnail = wc_get_image_size('gallery_thumbnail');
+        $thumbnail_size = apply_filters('woocommerce_gallery_thumbnail_size', [
+          $gallery_thumbnail['width'],
+          $gallery_thumbnail['height']
+        ]);
+        $thumbnail_url = wp_get_attachment_image_src($thumbnail_id, $thumbnail_size);
+        $thumbnail_url = isset($thumbnail_url[0]) ? $thumbnail_url[0] : '';
```

Comment:
What do you need the product thumbnail for if you are displaying a video?

Reply:
This `data-thumb` attribute that I'm filling with the `$thumbnail_url` is used by the default WooCommerce `flexslider.js`. I'm unsure though, whether `flexslider.js` is still responsible for the product images slider on product detail pages if Gallerya is in use – do you know that?